cloud-nio/cloud-nio-impl-drs/src/main/scala/cloud/nio/impl/drs/DrsCloudNioFileProvider.scala

```diff
@@ -2,64 +2,107 @@ package cloud.nio.impl.drs
 
 import java.nio.channels.{ReadableByteChannel, WritableByteChannel}
 
+import cats.effect.IO
 import cloud.nio.spi.{CloudNioFileList, CloudNioFileProvider, CloudNioRegularFileAttributes}
+import com.google.auth.oauth2.{AccessToken, OAuth2Credentials}
 import com.typesafe.config.Config
+import org.apache.commons.lang3.exception.ExceptionUtils
 import org.apache.http.HttpStatus
-import org.apache.http.client.methods.CloseableHttpResponse
-import org.apache.http.impl.client.{CloseableHttpClient, HttpClientBuilder}
+import org.apache.http.impl.client.HttpClientBuilder
 
-import scala.util.Try
+import scala.concurrent.duration._
 
 
-class DrsCloudNioFileProvider(config: Config, scheme: String, drsPathResolver: DrsPathResolver) extends CloudNioFileProvider {
+class DrsCloudNioFileProvider(config: Config,
+                              scheme: String,
+                              drsPathResolver: DrsPathResolver,
+                              authCredentials: OAuth2Credentials,
+                              httpClientBuilder: HttpClientBuilder,
+                              drsReadInterpreter: (MarthaResponse) => IO[ReadableByteChannel]) extends CloudNioFileProvider {
 
-  private lazy val httpClientBuilder = HttpClientBuilder.create()
+  private val AccessTokenAcceptableTTL = 1.minute
 
   private def getDrsPath(cloudHost: String, cloudPath: String): String = s"$scheme://$cloudHost/$cloudPath"
 
-  private def checkIfPathExistsThroughMartha(drsPath: String): Boolean = {
-    val httpClient: CloseableHttpClient = httpClientBuilder.build()
-
-    try {
-      val marthaResponse: CloseableHttpResponse = drsPathResolver.makeHttpRequestToMartha(drsPath, httpClient)
-
-      try {
-        marthaResponse.getStatusLine.getStatusCode == HttpStatus.SC_OK
-      } finally {
-        Try(marthaResponse.close())
-        ()
-      }
-    } finally {
-      Try(httpClient.close())
-      ()
+
+  private def getFreshAccessToken(credential: OAuth2Credentials): String = {
+    def accessTokenTTLIsAcceptable(accessToken: AccessToken): Boolean = {
+      (accessToken.getExpirationTime.getTime - System.currentTimeMillis()).millis.gteq(AccessTokenAcceptableTTL)
+    }
+
+    Option(credential.getAccessToken) match {
```

Comment:
ToL `cats.syntax.option` gives you `fold` which is nicer than pattern matching, when combined with consider calling `.filter(accessTokenTTLIsAcceptable)`